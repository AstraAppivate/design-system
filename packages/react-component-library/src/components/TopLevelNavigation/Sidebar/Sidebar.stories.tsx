import React from 'react'
import styled from 'styled-components'
import { Meta } from '@storybook/react/types-6-0'

import { Graph, House, Tools } from '../../../icons'
import { Link } from '../../Link'
import { Notification, Notifications } from '../NotificationPanel'
import { Sidebar, SidebarNav, SidebarNavItem, SidebarUser } from './index'

export default {
  component: Sidebar,
  subcomponents: {
    SidebarNav,
    SidebarNavItem,
    SidebarUser,
    Notifications,
    Notification,
  },
  title: 'Sidebar (Deprecated)',
} as Meta

const nav = (
  <SidebarNav>
    <SidebarNavItem Image={House} link={<Link href="#">Home</Link>} />
    <SidebarNavItem Image={Graph} link={<Link href="#">Stats</Link>} isActive />
    <SidebarNavItem Image={Tools} link={<Link href="#">Tools</Link>} />
  </SidebarNav>
)

const notifications = (
  <Notifications link={<Link href="#" />}>
    <Notification
      link={<Link href="#" />}
      name="Thomas Stephens"
      action="added a new comment to your"
      on="review"
      when={new Date('2019-11-05T14:25:02.178Z')}
      description="At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores"
    />
    <Notification
      link={<Link href="#" />}
      name="Thomas Stephens"
      action="added a new comment to your"
      on="review"
      when={new Date('2019-11-01T14:25:02.178Z')}
      description="At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores"
    />
    <Notification
      link={<Link href="#" />}
      name="Thomas Stephens"
      action="added a new comment to your"
      on="review"
      when={new Date('2019-11-01T14:25:02.178Z')}
      description="At vero eos et accusamus et iusto odio dignissimos ducimus qui blanditiis praesentium voluptatum deleniti atque corrupti quos dolores"
    />
  </Notifications>
)

const user = <SidebarUser initials="XT" link={<Link href="#" />} />

const StyledSidebar = styled(Sidebar)``

export const Default = (props: any) => <Sidebar {...props} nav={nav} />

Default.args = {}

export const WithNotifications = () => (
  <Sidebar hasUnreadNotification nav={nav} notifications={notifications} />
)

WithNotifications.storyName = 'With notifications'

export const WithUser = () => <Sidebar nav={nav} user={user} />

WithUser.storyName = 'With user'
