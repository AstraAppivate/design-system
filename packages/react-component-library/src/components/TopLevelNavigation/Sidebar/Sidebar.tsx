import React, { useState } from 'react'
import classNames from 'classnames'

import {
  NOTIFICATION_CONTAINER_WIDTH,
  NotificationsProps,
} from '../NotificationPanel'
import { SidebarNavProps, SidebarUserProps } from './index'
import { Sheet } from '../Sheet'
import { SheetButton } from '../SheetButton'
import { Bell } from '../../../icons'

interface SidebarProps {
  hasUnreadNotification?: boolean
  nav?: React.ReactElement<SidebarNavProps>
  notifications?: React.ReactElement<NotificationsProps>
  user?: React.ReactElement<SidebarUserProps>
}

export const Sidebar: React.FC<SidebarProps> = ({
  nav,
  notifications,
  hasUnreadNotification,
  user,
}) => {
  const [open, setOpen] = useState(false)
  const classes = classNames('rn-sidebar', {
    'is-open': open,
  })

  return (
    <div className={classes} data-testid="sidebar">
      {nav &&
        React.cloneElement(nav, {
          ...nav.props,
          onBlur: () => setOpen(false),
          onFocus: () => setOpen(true),
          onItemClick: () => setOpen(false),
          onMouseOut: () => setOpen(false),
          onMouseOver: () => setOpen(true),
        })}

      <div className="rn-sidebar__bottom">
        {notifications && (
          <Sheet
            button={(
              <SheetButton
                data-testid="notification-button"
                icon={<Bell className="rn-sheet__icon" />}
              >
                {hasUnreadNotification && (
                  <span
                    className="rn-notification-panel__not-read"
                    data-testid="not-read"
                  />
                )}
              </SheetButton>
            )}
            width={NOTIFICATION_CONTAINER_WIDTH}
          >
            {notifications}
          </Sheet>
        )}
        {user}
      </div>
    </div>
  )
}

Sidebar.displayName = 'Sidebar'
