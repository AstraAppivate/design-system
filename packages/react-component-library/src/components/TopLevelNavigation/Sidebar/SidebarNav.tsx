import React from 'react'

import { Nav, NavItem } from '../../../common/Nav'
import { SidebarNavItem, SidebarNavItemProps } from './index'
import { warnIfOverwriting } from '../../../helpers'

export interface SidebarNavProps extends Nav<NavItem> {
  /**
   * Optional handler invoked when `onBlur` event is emitted.
   */
  onBlur?: () => void
  /**
   * Optional handler invoked when `onFocus` event is emitted.
   */
  onFocus?: () => void
  /**
   * Optional handler invoked when an item is clicked.
   */
  onItemClick?: () => void
  /**
   * Optional handler invoked when the `onMouseOut` event is emitted.
   */
  onMouseOut?: () => void
  /**
   * Optional handler invoked when the `onMouseOver` event is emitted.
   */
  onMouseOver?: () => void
}

function mapNavItem(
  navItem: React.ReactElement<SidebarNavItemProps>,
  onClick: () => void
) {
  warnIfOverwriting(navItem.props, 'onClick', SidebarNavItem.name)

  return React.cloneElement(navItem, {
    ...navItem.props,
    onClick,
  })
}

/**
 * @deprecated
 */
export const SidebarNav: React.FC<SidebarNavProps> = ({
  children,
  onItemClick,
  ...rest
}) => (
  <nav className="rn-sidebar__top" data-testid="sidebar-nav" {...rest}>
    {React.Children.map(
      children,
      (child: React.ReactElement<SidebarNavItemProps>) => {
        return mapNavItem(child, onItemClick)
      }
    )}
  </nav>
)
