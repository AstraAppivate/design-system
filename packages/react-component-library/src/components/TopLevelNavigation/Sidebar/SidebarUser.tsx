import React, { ReactElement } from 'react'
import classNames from 'classnames'

import { Avatar } from '../../Avatar'
import { ComponentWithClass } from '../../../common/ComponentWithClass'
import { LinkTypes } from '../../../common/Link'

export interface SidebarUserProps extends ComponentWithClass {
  /**
   * Initials of the end user (e.g. Joe Bloggs => JB).
   */
  initials: string
  /**
   * Link component to apply to the user avatar.
   */
  link: React.ReactElement<LinkTypes>
}

/**
 * @deprecated
 */
export const SidebarUser: React.FC<SidebarUserProps> = ({
  className,
  initials,
  link,
  ...rest
}) => {
  const classes = classNames('rn-sidebar__nav-link', className)

  return React.cloneElement(link as ReactElement, {
    ...link.props,
    children: (
      <Avatar
        className="rn-sidebar__avatar rn-sidebar__nav-label"
        initials={initials}
      />
    ),
    className: classes,
    'data-testid': 'sidebar-user',
    ...rest,
  })
}

SidebarUser.displayName = 'SidebarUser'
