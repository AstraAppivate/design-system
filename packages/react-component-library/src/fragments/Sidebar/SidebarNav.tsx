import React from 'react'

import { Nav } from '../../types/Nav'
import { SidebarNavItemProps } from './SidebarNavItem'

export interface SidebarNavProps extends Nav {
  onBlur?: () => void
  onFocus?: () => void
  onItemClick?: () => void
  onMouseOut?: () => void
  onMouseOver?: () => void
}

function mapNavItem(
  navItem: React.ReactElement<SidebarNavItemProps>,
  onClick: () => void
) {
  if (navItem.props.onClick) {
    console.warn('Prop `onClick` on `SidebarNavItem` will be overwritten')
  }
  return React.cloneElement(navItem, {
    ...navItem.props,
    onClick,
  })
}

export const SidebarNav: React.FC<SidebarNavProps> = ({
  children,
  onBlur,
  onFocus,
  onItemClick,
  onMouseOut,
  onMouseOver,
}) => {
  const childrenAsArray = children as React.ReactElement<SidebarNavItemProps>[]

  return (
    <nav
      className="rn-sidebar__top"
      data-testid="sidebar-nav"
      onBlur={onBlur}
      onFocus={onFocus}
      onMouseOut={onMouseOut}
      onMouseOver={onMouseOver}
    >
      {React.Children.map(childrenAsArray, child => {
        return mapNavItem(child, onItemClick)
      })}
    </nav>
  )
}
