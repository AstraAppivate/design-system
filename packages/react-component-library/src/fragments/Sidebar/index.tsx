import React, { useState } from 'react'

import { Avatar, Link } from '../../components'
import NotificationSidePanel from '../NotificationSidePanel'

interface UserLinkType {
  initials: string
}

interface SidebarProps {
  navItems: NavItemTypes[]
  NotificationsPopoverContent?: JSX.Element
  unreadNotification?: boolean
  user?: UserLinkType
  LinkComponent?: any
}

const Sidebar: React.FC<SidebarProps> = ({
  LinkComponent = Link,
  navItems,
  NotificationsPopoverContent,
  unreadNotification = false,
  user,
}) => {
  const [open, setOpen] = useState()

  const renderUserLink = ({ initials, ...rest }: UserLinkType): JSX.Element => (
    <LinkComponent className="rn-sidebar__nav-link" {...rest}>
      <Avatar
        className="rn-sidebar__avatar rn-sidebar__nav-label"
        initials={initials}
      />
    </LinkComponent>
  )

  const renderSidebarNavItem = ({
    active = false,
    Image,
    label,
    ...rest
  }: NavItemTypes): JSX.Element => (
    <LinkComponent
      {...rest}
      key={label}
      className={`rn-sidebar__nav-link${active ? ' is-active' : ''}`}
    >
      <span className="rn-sidebar__nav-icon">{Image && <Image />}</span>
      <span className="rn-sidebar__nav-label">{label}</span>
    </LinkComponent>
  )

  const sideBarNavItemElements = navItems.map(renderSidebarNavItem)
  const userElement = user ? renderUserLink(user) : null

  return (
    <div className={`rn-sidebar ${open ? 'is-open' : 'is-closed'}`}>
      <nav
        className="rn-sidebar__top"
        onMouseOver={() => setOpen(true)}
        onFocus={() => setOpen(true)}
        onMouseOut={() => {
          setOpen(false)
        }}
        onBlur={() => {
          setOpen(false)
        }}
      >
        {sideBarNavItemElements}
      </nav>

      <div className="rn-sidebar__bottom">
        {NotificationsPopoverContent && (
          <NotificationSidePanel
            className="rn-sidebar__nav-link"
            unreadNotification={unreadNotification}
          >
            {NotificationsPopoverContent}
          </NotificationSidePanel>
        )}
        {userElement}
      </div>
    </div>
  )
}

export default Sidebar
