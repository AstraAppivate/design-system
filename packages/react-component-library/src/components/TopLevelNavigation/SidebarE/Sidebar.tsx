import React from 'react'
import styled from 'styled-components'
import { selectors } from '@royalnavy/design-tokens'
import { Transition } from 'react-transition-group'

import { SidebarHandle } from './SidebarHandle'
import { SidebarUserEProps } from './SidebarUser'
import { SidebarNotifications } from './SidebarNotifications'
import { ComponentWithClass } from '../../../common/ComponentWithClass'
import { SidebarContext, SidebarProvider } from './context'
import { NotificationsProps } from '../NotificationPanel'
import { TRANSITION_STYLES, TRANSITION_TIMEOUT } from './constants'

export interface SidebarEProps extends ComponentWithClass {
  icon?: React.ReactNode
  title?: string
  user?: React.ReactElement<SidebarUserEProps>
  hasUnreadNotification?: boolean
  notifications?: React.ReactElement<NotificationsProps>
}

interface StyledSidebarProps {
  isOpen?: boolean
}

const { color, spacing, fontSize } = selectors

const StyledSidebar = styled.aside<StyledSidebarProps>`
  display: inline-flex;
  flex-direction: column;
  position: relative;
  width: ${({ isOpen }) => (isOpen ? '18rem' : '3.75rem')};
  height: 100vh;
  background-color: ${color('neutral', '700')};
  color: ${color('neutral', 'white')};
  transition: ${({ isOpen }) =>
    isOpen
      ? '200ms width cubic-bezier(0.34, 1.56, 0.64, 1)'
      : '200ms width cubic-bezier(0.34, 1, 0.64, 1)'};

  > a:hover {
    text-decoration: none;
  }
`

const StyledHead = styled.div`
  display: flex;
  align-items: center;
  padding: ${spacing('8')} ${spacing('6')};
  background-color: ${color('neutral', '800')};
`

const StyledIcon = styled.div`
  display: inline-flex;
  align-items: center;
  border-radius: 4px;
  background-color: ${color('neutral', '500')};
  padding: 0.55rem;
  margin-right: ${spacing('7')};

  svg {
    width: 18px;
    height: 18px;
  }
`

const StyledTitle = styled.div`
  font-weight: 600;
  font-size: ${fontSize('l')};
  white-space: nowrap;
  opacity: 1;
  transition: opacity 150ms linear;
`

export const SidebarE: React.FC<SidebarEProps> = ({
  icon,
  title,
  children,
  user,
  hasUnreadNotification,
  notifications,
}) => {
  return (
    <SidebarProvider>
      <SidebarContext.Consumer>
        {({ isOpen, hasMouseOver, setHasMouseOver }) => (
          <StyledSidebar
            isOpen={isOpen}
            onMouseEnter={(_) => setHasMouseOver(true)}
            onMouseLeave={(_) => setHasMouseOver(false)}
          >
            <Transition in={hasMouseOver} timeout={0} unmountOnExit>
              {(state) => (
                <SidebarHandle style={{ ...TRANSITION_STYLES[state] }} />
              )}
            </Transition>
            {title && (
              <StyledHead data-testid="sidebar-head">
                {icon && <StyledIcon>{icon}</StyledIcon>}
                <Transition
                  in={isOpen}
                  timeout={TRANSITION_TIMEOUT}
                  unmountOnExit
                >
                  {(state) => (
                    <StyledTitle style={{ ...TRANSITION_STYLES[state] }}>
                      {title}
                    </StyledTitle>
                  )}
                </Transition>
              </StyledHead>
            )}
            {children}

            {notifications && (
              <SidebarNotifications
                notifications={notifications}
                hasUnreadNotification={hasUnreadNotification}
              />
            )}

            {user}
          </StyledSidebar>
        )}
      </SidebarContext.Consumer>
    </SidebarProvider>
  )
}

SidebarE.displayName = 'Sidebar'
