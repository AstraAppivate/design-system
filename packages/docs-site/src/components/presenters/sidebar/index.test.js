import React from 'react'
import renderer from 'react-test-renderer'

import Sidebar from './index'

/* eslint-disable react/no-children-prop */

const navigation = [
  {
    node: {
      id: 'bc54746a-2d34-534b-bdd8-86c1d13d95bb',
      fields: {
        slug: '/',
      },
      frontmatter: {
        title: 'Home',
        status: '',
        index: 0,
      },
    },
    children: [],
  },
]

describe('Sidebar', () => {
  it('snapshot: has same HTML structure', () => {
    const tree = renderer.create(<Sidebar navigation={navigation} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
