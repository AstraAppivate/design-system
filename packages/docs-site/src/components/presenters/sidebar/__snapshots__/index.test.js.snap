// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar snapshot: has same HTML structure 1`] = `
<nav>
  <ul>
    <li>
      <a
        href="/"
      >
        <span>
          Home
        </span>
      </a>
    </li>
  </ul>
  <pre>
    [
  {
    "node": {
      "id": "bc54746a-2d34-534b-bdd8-86c1d13d95bb",
      "fields": {
        "slug": "/"
      },
      "frontmatter": {
        "title": "Home",
        "status": "",
        "index": 0
      }
    },
    "children": []
  }
]
  </pre>
</nav>
`;
